stac_mjx/controller.py

```diff
@@ -1,330 +1,546 @@
+            get_batch_offsets = jax.vmap(op.get_site_pos, in_axes=(0, None))
+            offsets = get_batch_offsets(mjx_model, self._body_site_idxs).copy()[0]
+            x = x.reshape(-1, x.shape[-1])
+            q = q.reshape(-1, q.shape[-1])
+        else:
+            offsets = op.get_site_pos(mjx_model, self._body_site_idxs).copy()
+
+        kp_data = kp_data.reshape(-1, kp_data.shape[-1])
+
+        data = {}
+
+        for k, v in self.model_cfg.items():
+            data[k] = v
+
+        data.update(
+            {
+                "qpos": q,
+                "xpos": x,
+                "walker_body_sites": walker_body_sites,
+                "offsets": offsets,
+                "names_qpos": self._part_names,
+                "names_xpos": self._body_names,
+                "kp_data": jp.copy(kp_data),
+                "kp_names": self._kp_names,
+            }
+        )
 
-        # Set the offsets.
-        mjx_model = op.set_site_pos(mjx_model, offsets)
+        return data
 
-        # forward is used to calculate xpos and such
-        mjx_data = mjx.kinematics(mjx_model, mjx_data)
-        mjx_data = mjx.com_pos(mjx_model, mjx_data)
+    def _create_keypoint_sites(self):
+        """Create sites for keypoints (used for rendering only).
 
-        return mjx_model, mjx_data
+        Returns:
+            (mujoco.Model, List, List): Mj_model for rendering, list of keypoint site indices, and list of body site indices
+        """
+        keypoint_sites = []
+        keypoint_site_names = []
+        # set up keypoint rendering by adding the kp sites to the root body
+        for id, name in enumerate(self.model_cfg["KEYPOINT_MODEL_PAIRS"]):
+            start = (np.random.rand(3) - 0.5) * 0.001
+            rgba = self.model_cfg["KEYPOINT_COLOR_PAIRS"][name]
+            site_name = name + "_kp"
+            keypoint_site_names.append(site_name)
+            site = self._root.worldbody.add(
+                "site",
+                name=site_name,
+                type="sphere",
+                size=[0.005],
+                rgba=rgba,
+                pos=start,
+                group=2,
+            )
+            keypoint_sites.append(site)
+
+        physics = mjcf.Physics.from_mjcf_model(self._root)
+
+        axis = physics.named.model.site_pos._axes[0]
+        # Combine the two lists of site names and create the index map
+        site_index_map = {
+            key: int(axis.convert_key_item(key))
+            for key in list(self.model_cfg["KEYPOINT_MODEL_PAIRS"].keys())
+            + keypoint_site_names
+        }
+        body_site_idxs = [
+            site_index_map[n] for n in self.model_cfg["KEYPOINT_MODEL_PAIRS"].keys()
+        ]
+        keypoint_site_idxs = [site_index_map[n] for n in keypoint_site_names]
+        self._body_site_idxs = body_site_idxs
+        self._keypoint_site_idxs = keypoint_site_idxs
+
+        return deepcopy(physics.model.ptr), body_site_idxs, keypoint_site_idxs
+
+    def render(
```

Comment:
I moved it here because the model with body sites is initialized here, along with the (same) parameters needed to create the keypoint sites. I couldn't quickly think of a way to move it out cleanly, without doing a lot of stuff that is more easily handled by this class, like updating the keypoint site positions. (The init function of this class doesn't do too much else other than what rendering would want as well).
Let me know know what you think! The concerns with this choice are likely similar to the overall ones you mentioned above, which I'm sympathetic to